src/Cryptor.ts

```diff
@@ -1,131 +1,160 @@
-		this._algo = algo;
-    	this._initVector = initVector ? initVector : this.genRandomIVSync();
+	private _algorithm: string
+	private _initVector: Buffer
+	private _key: Buffer
+	private _cipher: crypto.Cipher
+	private _decipher: crypto.Decipher
+	private _useWebWorkers: boolean
+	private _cryptorWorker?: ModuleThread<CryptorWorker>
+	constructor(
+		key: Buffer | string,
+		initVector: Buffer = crypto.randomBytes(16),
+		algorithm: string = 'aes-256-gcm',
+		useWebWorkers: boolean = false
+	) {
+		this._algorithm = algorithm
+		this._initVector = initVector
 		// TODO: generate salt ?
-		this._key = this._pbkdfSync(pass);
-		this._cipher = crypto.createCipheriv(algo, this._key, this._initVector);
-		this._decipher = crypto.createDecipheriv(algo, this._key, this._initVector);
+		this._key = this._pbkdfSync(key)
+		this._cipher = crypto.createCipheriv(algorithm, this._key, this._initVector)
+		this._decipher = crypto.createDecipheriv(algorithm, this._key, this._initVector)
+		// Async via Process or Web workers
+		this._useWebWorkers = useWebWorkers
+		if (this._useWebWorkers) {
+			spawn<CryptorWorker>(new Worker("./CryptorWorker")).then((worker) => {
+				this._cryptorWorker = worker
+				this._cryptorWorker.init(this._algorithm, this._key, this._initVector)
+			})
+		}
 	}
 
 
-	encryptSync(plainBuf: crypto.BinaryLike, initVector=null) {
+	encryptSync(plainBuf: crypto.BinaryLike, initVector?: Buffer): Buffer {
 		if (initVector && (initVector !== this._initVector)) {
-			this._resetCipher(initVector);
+			this._resetCipherSync(initVector!)
 		}
-		return this._cipher.update(plainBuf);
+		return this._cipher.update(plainBuf)
 	}
 
 	// TODO: needs iv param
-	encrypt(...args: Array<any>) {
-		console.log(this._key.toString('hex'));
-		console.log(this._initVector.toString('hex'));
-		let argSplit = this._separateCallback(args)
-		let cb = argSplit.cb;
-		let methodArgs = argSplit.args;
-
-
-		this._callAsync(
-			this.encryptSync.bind(this),
-			methodArgs,
-			cb
-		);
-		return;
-	}
-
-	decryptSync(cipherBuf, initVector=null) {
+	async encrypt(plainBuf: crypto.BinaryLike, initVector: Buffer | undefined = undefined, callback: (err: NodeJS.ErrnoException | null, buffer: Buffer) => void): Promise<void> {
 		if (initVector && (initVector !== this._initVector)) {
-			this._resetDecipher(initVector);
+			this._resetCipher(initVector!)
 		}
 
-		return this._decipher.update(cipherBuf);
+		let buffer: Buffer
+		if (this._useWebWorkers && this._cryptorWorker) {
+			buffer = await this._cryptorWorker.updateCipher(this._algorithm, this._key, this._initVector, plainBuf)
+		} else {
+			buffer = this._cipher.update(plainBuf)
+		}
+		callback(null, buffer)
 	}
 
-	// TODO: needs iv param
-	decrypt(...args: Array<any>) {
-		let argSplit = this._separateCallback(args)
-		let cb = argSplit.cb;
-		let methodArgs = argSplit.args;
+	decryptSync(cipherBuf: NodeJS.ArrayBufferView, initVector?: Buffer): Buffer {
```

Comment:
Could you explain what does `NodeJS.ArrayBufferView` come from? I was under the assumption that native JS types which `ArrayBufferView` is is just global in TS. So you can just say `ArrayBufferView`.

And if you do use `NodeJS` what does this mean for browsers or non-nodejs environments?

Reply:
This was just an initial attempt on my behalf to satisfy typescript in providing a type that feeds into the crypto library but we should be using the more relevant `Buffer` class

Reply:
this also applies to the lines where crypto.BinaryLike is specified and can be replaced with `Buffer | string`